### PR TITLE
Fuse module enhancements

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26457 Fuse module enhancements**

Enhancement to fuse module to support sequentials, fuse list can now be just like the state dict.
Also add support for Conv-Relu and linear-relu fusion

Differential Revision: [D17466382](https://our.internmc.facebook.com/intern/diff/D17466382/)

### CODE REVIEW

File: torch/nn/_intrinsic/qat/modules/conv_fused.py [hunk 1]
```diff
@@ -246,17 +246,16 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                  qconfig=None):
         super(ConvReLU2d, self).__init__(in_channels, out_channels, kernel_size,
                                          stride=stride, padding=padding, dilation=dilation,
-                                         groups=groups, bias=bias, padding_mode=padding_mode)
+                                         groups=groups, bias=bias, padding_mode=padding_mode,
+                                         qconfig = qconfig)
```
Comment (dzhulgakov) [inline torch/nn/_intrinsic/qat/modules/conv_fused.py]: nit: formatting

File: torch/quantization/fuse_modules.py [hunk 2]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
         return torch_fused.ConvReLU2d(
             torch.nn.utils.fusion.fuse_conv_bn_eval(conv, bn), relu)
 
+def _get_module(model, submodule_key):
+    submodule = ''
+    tokens = submodule_key.split('.')
+    cur_mod = model
+    for s in tokens:
+        if s.isdigit():
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: I just tried and I don't think you need this - getattr seems to work just fine on string keys that are numbers

File: torch/quantization/fuse_modules.py [hunk 3]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
+
+# Generalization of setattr
+def _set_module(model, submodule_key, module):
+    submodule = ''
+    tokens = submodule_key.split('.')
+    sub_tokens = tokens[:-1]
+    cur_mod = model
+    prev_mod = model
+    for s in sub_tokens:
+        if s.isdigit():
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: same

File: torch/quantization/fuse_modules.py [hunk 4]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
+            cur_mod = cur_mod[int(s)]
+        else:
+            cur_mod = getattr( cur_mod, s)
+
+    if tokens[-1].isdigit():
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: same

File: torch/quantization/fuse_modules.py [hunk 5]
```diff
@@ -3,6 +3,15 @@
...
+    r"""Given the conv and relu modules, fuses them and returns the fused module
+    """
+    return torch.nn._intrinsic.ConvReLU2d(conv, relu)
+
+def fuse_linear_relu(linear, relu):
+    r"""Given the linear and relu modules, fuses them and returns the fused module
+    """
+    return torch.nn._intrinsic.LinearReLU(linear, relu)
 
 def fuse_conv_bn(conv, bn):
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: this function (and probably `fuse_conv_bn_eval`) misses setting the training mode. Something like `result.train(original.training)`

File: torch/quantization/fuse_modules.py [hunk 6]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
-        parent_module_name = '.'.join(modules_to_fuse[i].split('.')[:-1])
-        mod.append(named_module_dict[modules_to_fuse[i]])
-        parent_mod.append(named_module_dict.get(parent_module_name, model))
+    for item in modules_to_fuse:
+        mod.append(_get_module(model, item))
 
-    new_mod = mod[0]
     if fuser_func is None:
         types = tuple(type(m) for m in mod)
         fuser_func = OP_LIST_TO_FUSER_FUNC.get(types, None)
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: I'd pull this logic into something like 'fuse_known_modules' and pass it as a default to `fuser_func` - it's clear this way and makes it easier to extend functionality

File: torch/quantization/fuse_modules.py [hunk 7]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
 
-    # Assign new_mod to module and set remaining modules to identity
-    if new_mod is not mod[0]:
-        setattr(parent_mod[0], modules_to_fuse[0].split('.')[-1], new_mod)
-        for i in range(1, len(modules_to_fuse)):
-            setattr(parent_mod[i], modules_to_fuse[i].split('.')[-1], torch.nn.Identity())
+    _set_module(model, modules_to_fuse[0], new_mod)
 
+    for item in modules_to_fuse[1:]:
+        _set_module(model, item, torch.nn.Identity())
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: for cleanness - set the training mode here too

File: torch/quantization/fuse_modules.py [hunk 8]
```diff
@@ -111,6 +144,5 @@ def fuse_modules(model, modules_to_fuse):
             >>> output = m(input)
 
     """
-    named_module_dict = {name: mod for name, mod in model.named_modules()}
     for module_list in modules_to_fuse:
-        _fuse_modules(model, named_module_dict, module_list)
+        _fuse_modules(model, module_list)
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: can we add a check that if passed in thing is a list of strings just do it? (and update the doc accordingly)

```
if len(modules_to_fuse) > 0 and isinstance(modules_to_fuse, str):
  _fuse_modules(model, modules_to_fuse)
else:
  for module_list ...
```

File: torch/quantization/fuse_modules.py [hunk 9]
```diff
@@ -52,44 +61,68 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
-    if new_mod is not mod[0]:
-        setattr(parent_mod[0], modules_to_fuse[0].split('.')[-1], new_mod)
-        for i in range(1, len(modules_to_fuse)):
-            setattr(parent_mod[i], modules_to_fuse[i].split('.')[-1], torch.nn.Identity())
+    _set_module(model, modules_to_fuse[0], new_mod)
 
+    for item in modules_to_fuse[1:]:
+        _set_module(model, item, torch.nn.Identity())
 
 def fuse_modules(model, modules_to_fuse):
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: I wonder whether we should do special handling for Sequential. The good thing about sequantials is that we **know** what forward looks like. So we could just iterate through them and try to find a match.

I'm also fine doing it later in a separate PR as it doesn't appear to be a critical feature.

**Review by dzhulgakov — COMMENTED** [review]
Looking good, some more suggestions

File: test/test_quantization.py [hunk 10]
```diff
@@ -14,7 +14,7 @@
 from common_utils import run_tests
 from common_quantization import QuantizationTestCase, SingleLayerLinearModel, \
     SkipQuantModel, QuantStubModel, \
-    ModelForFusion, ManualLinearQATModel, ManualConvLinearQATModel, \
+    ModelForFusion, ModelForFusion2, ManualLinearQATModel, ManualConvLinearQATModel, \
```
Comment (jerryzh168) [inline test/test_quantization.py]: nit: can you rename ModelForFusion and ModelForFusion2 to more informative names

File: torch/quantization/fuse_modules.py (hunk 5 above)
Comment (raghuramank100) [reply torch/quantization/fuse_modules.py]: There is no need to do that here, we just create a sequential here with the original modules, so all attributes are preserved

File: torch/quantization/fuse_modules.py (hunk 8 above)
Comment (raghuramank100) [reply torch/quantization/fuse_modules.py]: What is passed in is a list of list of strings, like, for example:
```python
        fuse_modules(model, [['conv1', 'relu1'] ,
                             ['features.0.0', 'features.0.1', 'features.0.2'],
                             ['features.1.0', 'features.1.1', 'features.1.2'],
                             ['features.2.0', 'features.2.1', 'features.2.2'],
                             ['classifier.0', 'classifier.1']])
```
So, not sure what you mean here? can you elaborate?

File: torch/quantization/fuse_modules.py (hunk 9 above)
Comment (raghuramank100) [reply torch/quantization/fuse_modules.py]: This is useful only when we have a sequential module at the leaf of the module tree having fusable modules, so lets do this in a later PR

File: torch/quantization/fuse_modules.py [hunk 11]
```diff
@@ -1,8 +1,18 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import torch
+import copy
 
 import torch.nn._intrinsic.modules.fused as torch_fused
+def fuse_conv_relu(conv, relu):
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: you don't need these function - just put torch.nn._intrinsic.ConvReLU2d in the dict below

File: torch/quantization/fuse_modules.py (hunk 8 above)
Comment (dzhulgakov) [reply torch/quantization/fuse_modules.py]: I mean also supporting passing a single list of strings (in addition to list of list of strings). It's just convenient to also accept it without extra nesting:

`fuse_modules(model, ['conv1', 'relu1'])`

File: torch/quantization/fuse_modules.py [hunk 12]
```diff
@@ -52,44 +62,78 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
+        mod_list.append(_get_module(model, item))
+
+    # Fuse list of modules
+    new_mod_list = fuser_func(mod_list)
+
+    # Replace original module list with fused module list
+    for i, item in enumerate(modules_to_fuse):
+        _set_module(model, item, new_mod_list[i])
+
+def fuse_modules(model, modules_to_fuse, inplace=False):
```
Comment (dzhulgakov) [inline torch/quantization/fuse_modules.py]: you probably wanted to expose fuser_func here too?

**Review by dzhulgakov — APPROVED** [review]
Minor comments - otherwise good to go
